File: components/monitoring/sdk/r/R/monitoring.R
#' Wafer Monitor R SDK
#'
#' @description
#' R client library for monitoring distributed applications.
#'
#' @section Main Functions:
#' \itemize{
#'   \item{\code{\link{monitoring_init}}: Initialize the SDK}
#'   \item{\code{\link{monitoring_start}}: Start a monitored context}
#'   \item{\code{\link{monitoring_progress}}: Report progress}
#'   \item{\code{\link{monitoring_finish}}: Finish successfully}
#'   \item{\code{\link{monitoring_error}}: Finish with error}
#' }
#'
#' @docType package
#' @name monitoring-package
NULL

# Global SDK state
.sdk_state <- new.env(parent = emptyenv())
.sdk_state$initialized <- FALSE
.sdk_state$config <- NULL
.sdk_state$backend <- NULL

#' Initialize Monitoring SDK
#'
#' @param mode Character. Either "sidecar" or "direct"
#' @param app_name Character. Name of the application
#' @param app_version Character. Version of the application
#' @param site_id Character. Site identifier
#' @param instance_id Character. Instance identifier (optional, defaults to hostname)
#' @param sidecar_url Character. URL of sidecar agent (for sidecar mode)
#' @param timeout Numeric. Request timeout in seconds (default: 5.0)
#' @param max_retries Integer. Maximum retries for failed requests (default: 3)
#' @param backend_type Character. Backend type for direct mode ("filesystem", "s3", etc.)
#' @param backend_config List. Backend-specific configuration
#'
#' @return Invisible NULL. Throws error on failure.
#'
#' @export
#' @examples
#' \dontrun{
#' monitoring_init(
#'   mode = "sidecar",
#'   app_name = "my-r-app",
#'   app_version = "1.0.0",
#'   site_id = "fab1",
#'   sidecar_url = "http://localhost:17000"
#' )
#' }
monitoring_init <- function(mode = c("sidecar", "direct"),
                           app_name,
                           app_version,
                           site_id,
                           instance_id = NULL,
                           sidecar_url = "http://localhost:17000",
                           timeout = 5.0,
                           max_retries = 3,
                           backend_type = "filesystem",
                           backend_config = list()) {
  
  if (.sdk_state$initialized) {
    stop("SDK already initialized. Call monitoring_shutdown() first.")
  }
  
  mode <- match.arg(mode)
  
  # Default instance_id to hostname
  if (is.null(instance_id)) {
    instance_id <- Sys.info()["nodename"]
  }
  
  # Store configuration
  .sdk_state$config <- list(
    mode = mode,
    app_name = app_name,
    app_version = app_version,
    site_id = site_id,
    instance_id = instance_id,
    sidecar_url = sidecar_url,
    timeout = timeout,
    max_retries = max_retries,
    backend_type = backend_type,
    backend_config = backend_config
  )
  
  # Initialize backend
  if (mode == "sidecar") {
    .sdk_state$backend <- SidecarBackend$new(
      url = sidecar_url,
      timeout = timeout,
      max_retries = max_retries
    )
  } else {
    if (backend_type == "filesystem") {
      .sdk_state$backend <- FilesystemBackend$new(backend_config)
    } else {
      stop(sprintf("Unsupported backend type: %s", backend_type))
    }
  }
  
  .sdk_state$initialized <- TRUE
  message(sprintf("✓ Monitoring SDK initialized (mode=%s)", mode))
  
  invisible(NULL)
}

#' Shutdown Monitoring SDK
#'
#' @return Invisible NULL
#' @export
monitoring_shutdown <- function() {
  if (!.sdk_state$initialized) {
    warning("SDK not initialized")
    return(invisible(NULL))
  }
  
  .sdk_state$initialized <- FALSE
  .sdk_state$config <- NULL
  .sdk_state$backend <- NULL
  
  message("✓ Monitoring SDK shut down")
  invisible(NULL)
}

#' Check if SDK is Initialized
#'
#' @return Logical. TRUE if initialized, FALSE otherwise
#' @export
monitoring_is_initialized <- function() {
  return(.sdk_state$initialized)
}

#' Start a Monitored Context
#'
#' @param name Character. Name of the job/task
#' @param entity_id Character. Unique identifier for the entity
#' @param entity_type Character. Type of entity ("job", "task", "step")
#'
#' @return A MonitoringContext R6 object
#' @export
#' @examples
#' \dontrun{
#' ctx <- monitoring_start("process-wafer", "W-12345")
#' monitoring_progress(ctx, 50, "Halfway done")
#' monitoring_finish(ctx)
#' }
monitoring_start <- function(name, entity_id, entity_type = "job") {
  if (!.sdk_state$initialized) {
    stop("SDK not initialized. Call monitoring_init() first.")
  }
  
  ctx <- MonitoringContext$new(
    name = name,
    entity_id = entity_id,
    entity_type = entity_type,
    config = .sdk_state$config,
    backend = .sdk_state$backend
  )
  
  ctx$start()
  return(ctx)
}

#' Report Progress
#'
#' @param ctx MonitoringContext object
#' @param progress Numeric. Progress percentage (0-100)
#' @param message Character. Status message
#'
#' @return Invisible NULL
#' @export
monitoring_progress <- function(ctx, progress, message = NULL) {
  if (!inherits(ctx, "MonitoringContext")) {
    stop("Invalid context object")
  }
  ctx$progress(progress, message)
  invisible(NULL)
}

#' Add Metric to Context
#'
#' @param ctx MonitoringContext object
#' @param key Character. Metric name
#' @param value Numeric. Metric value
#'
#' @return Invisible NULL
#' @export
monitoring_add_metric <- function(ctx, key, value) {
  if (!inherits(ctx, "MonitoringContext")) {
    stop("Invalid context object")
  }
  ctx$add_metric(key, value)
  invisible(NULL)
}

#' Add Metadata to Context
#'
#' @param ctx MonitoringContext object
#' @param key Character. Metadata key
#' @param value Character. Metadata value
#'
#' @return Invisible NULL
#' @export
monitoring_add_metadata <- function(ctx, key, value) {
  if (!inherits(ctx, "MonitoringContext")) {
    stop("Invalid context object")
  }
  ctx$add_metadata(key, value)
  invisible(NULL)
}

#' Finish Context Successfully
#'
#' @param ctx MonitoringContext object
#'
#' @return Invisible NULL
#' @export
monitoring_finish <- function(ctx) {
  if (!inherits(ctx, "MonitoringContext")) {
    stop("Invalid context object")
  }
  ctx$finish()
  invisible(NULL)
}

#' Finish Context with Error
#'
#' @param ctx MonitoringContext object
#' @param error_message Character. Error message
#'
#' @return Invisible NULL
#' @export
monitoring_error <- function(ctx, error_message) {
  if (!inherits(ctx, "MonitoringContext")) {
    stop("Invalid context object")
  }
  ctx$error(error_message)
  invisible(NULL)
}

#' Cancel Context
#'
#' @param ctx MonitoringContext object
#'
#' @return Invisible NULL
#' @export
monitoring_cancel <- function(ctx) {
  if (!inherits(ctx, "MonitoringContext")) {
    stop("Invalid context object")
  }
  ctx$cancel()
  invisible(NULL)
}

# ============================================================================
# MonitoringContext R6 Class
# ============================================================================

#' @import R6
MonitoringContext <- R6::R6Class(
  "MonitoringContext",
  
  public = list(
    name = NULL,
    entity_id = NULL,
    entity_type = NULL,
    config = NULL,
    backend = NULL,
    start_time = NULL,
    metrics = NULL,
    metadata = NULL,
    
    initialize = function(name, entity_id, entity_type, config, backend) {
      self$name <- name
      self$entity_id <- entity_id
      self$entity_type <- entity_type
      self$config <- config
      self$backend <- backend
      self$metrics <- list()
      self$metadata <- list()
    },
    
    start = function() {
      self$start_time <- Sys.time()
      
      event <- list(
        idempotency_key = sprintf("%s-start-%d", self$entity_id, as.numeric(self$start_time)),
        site_id = self$config$site_id,
        app_name = self$config$app_name,
        app_version = self$config$app_version,
        entity_type = self$entity_type,
        entity_id = self$entity_id,
        event_kind = "started",
        timestamp = as.numeric(self$start_time),
        status = "started",
        metrics = list(),
        metadata = list()
      )
      
      self$backend$send_event(event)
    },
    
    progress = function(progress, message = NULL) {
      event <- list(
        idempotency_key = sprintf("%s-progress-%d", self$entity_id, as.numeric(Sys.time())),
        site_id = self$config$site_id,
        app_name = self$config$app_name,
        app_version = self$config$app_version,
        entity_type = self$entity_type,
        entity_id = self$entity_id,
        event_kind = "progress",
        timestamp = as.numeric(Sys.time()),
        status = if (!is.null(message)) message else "in_progress",
        metrics = list(progress = progress),
        metadata = list()
      )
      
      self$backend$send_event(event)
    },
    
    add_metric = function(key, value) {
      self$metrics[[key]] <- value
    },
    
    add_metadata = function(key, value) {
      self$metadata[[key]] <- value
    },
    
    finish = function() {
      # Add duration metric
      duration <- as.numeric(difftime(Sys.time(), self$start_time, units = "secs"))
      self$metrics[["duration_seconds"]] <- duration
      
      event <- list(
        idempotency_key = sprintf("%s-finish-%d", self$entity_id, as.numeric(Sys.time())),
        site_id = self$config$site_id,
        app_name = self$config$app_name,
        app_version = self$config$app_version,
        entity_type = self$entity_type,
        entity_id = self$entity_id,
        event_kind = "finished",
        timestamp = as.numeric(Sys.time()),
        status = "success",
        metrics = self$metrics,
        metadata = self$metadata
      )
      
      self$backend$send_event(event)
    },
    
    error = function(error_message) {
      self$metadata[["error"]] <- error_message
      
      event <- list(
        idempotency_key = sprintf("%s-error-%d", self$entity_id, as.numeric(Sys.time())),
        site_id = self$config$site_id,
        app_name = self$config$app_name,
        app_version = self$config$app_version,
        entity_type = self$entity_type,
        entity_id = self$entity_id,
        event_kind = "error",
        timestamp = as.numeric(Sys.time()),
        status = "error",
        metrics = self$metrics,
        metadata = self$metadata
      )
      
      self$backend$send_event(event)
    },
    
    cancel = function() {
      event <- list(
        idempotency_key = sprintf("%s-cancel-%d", self$entity_id, as.numeric(Sys.time())),
        site_id = self$config$site_id,
        app_name = self$config$app_name,
        app_version = self$config$app_version,
        entity_type = self$entity_type,
        entity_id = self$entity_id,
        event_kind = "canceled",
        timestamp = as.numeric(Sys.time()),
        status = "canceled",
        metrics = list(),
        metadata = list()
      )
      
      self$backend$send_event(event)
    }
  )
)

# ============================================================================
# Backend Classes
# ============================================================================

#' @import R6
#' @import httr
#' @import jsonlite
SidecarBackend <- R6::R6Class(
  "SidecarBackend",
  
  public = list(
    url = NULL,
    timeout = NULL,
    max_retries = NULL,
    
    initialize = function(url, timeout, max_retries) {
      self$url <- url
      self$timeout <- timeout
      self$max_retries <- max_retries
    },
    
    send_event = function(event) {
      endpoint <- sprintf("%s/v1/ingest/events", self$url)
      
      for (attempt in 1:self$max_retries) {
        tryCatch({
          response <- httr::POST(
            endpoint,
            body = jsonlite::toJSON(event, auto_unbox = TRUE),
            httr::content_type_json(),
            httr::timeout(self$timeout)
          )
          
          if (httr::status_code(response) >= 200 && httr::status_code(response) < 300) {
            return(invisible(NULL))
          }
          
          if (httr::status_code(response) >= 400 && httr::status_code(response) < 500) {
            # Client error, don't retry
            warning(sprintf("HTTP error %d: %s", httr::status_code(response), httr::content(response, "text")))
            return(invisible(NULL))
          }
          
        }, error = function(e) {
          if (attempt == self$max_retries) {
            warning(sprintf("Failed to send event after %d attempts: %s", self$max_retries, e$message))
          } else {
            Sys.sleep(0.1 * (2^attempt))  # Exponential backoff
          }
        })
      }
      
      invisible(NULL)
    }
  )
)

#' @import R6
FilesystemBackend <- R6::R6Class(
  "FilesystemBackend",
  
  public = list(
    output_dir = NULL,
    filename_prefix = NULL,
    
    initialize = function(config) {
      self$output_dir <- config$output_dir %||% "./monitoring_events"
      self$filename_prefix <- config$filename_prefix %||% "events"
      
      # Create directory if it doesn't exist
      if (!dir.exists(self$output_dir)) {
        dir.create(self$output_dir, recursive = TRUE)
      }
    },
    
    send_event = function(event) {
      timestamp <- as.numeric(Sys.time())
      filename <- sprintf("%s/%s_%d_%d.jsonl", 
                         self$output_dir, 
                         self$filename_prefix,
                         timestamp,
                         Sys.getpid())
      
      json_line <- jsonlite::toJSON(event, auto_unbox = TRUE)
      
      tryCatch({
        write(json_line, file = filename, append = TRUE)
      }, error = function(e) {
        warning(sprintf("Failed to write event to file: %s", e$message))
      })
      
      invisible(NULL)
    }
  )
)

# Helper: NULL coalescing operator
`%||%` <- function(x, y) {
  if (is.null(x)) y else x
}

File: components/monitoring/sdk/c/src/monitoring.c
/**
 * @file monitoring.c
 * @brief Wafer Monitor C SDK - Core Implementation
 */

#include "monitoring.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <pthread.h>

#ifdef __APPLE__
#include <uuid/uuid.h>
#else
#include <uuid.h>
#endif

/* ========================================================================== */
/* Internal Structures                                                        */
/* ========================================================================== */

typedef struct backend_s {
    monitoring_backend_type_t type;
    bool enabled;
    int priority;
    void* impl;  /* Backend-specific implementation */
} backend_t;

typedef struct {
    bool initialized;
    monitoring_config_t config;
    backend_t* backends;
    int num_backends;
    pthread_mutex_t lock;
} sdk_state_t;

typedef struct monitoring_context_s {
    char* name;
    char* entity_id;
    monitoring_entity_type_t entity_type;
    time_t start_time;
    
    /* Dynamic arrays */
    char** metric_keys;
    double* metric_values;
    int num_metrics;
    int metrics_capacity;
    
    char** metadata_keys;
    char** metadata_values;
    int num_metadata;
    int metadata_capacity;
} monitoring_context_s;

/* Global SDK state */
static sdk_state_t g_sdk = {
    .initialized = false,
    .backends = NULL,
    .num_backends = 0,
    .lock = PTHREAD_MUTEX_INITIALIZER
};

/* ========================================================================== */
/* Forward Declarations                                                       */
/* ========================================================================== */

static monitoring_error_t send_event_to_backends(const monitoring_event_t* event);
static void free_context_internal(monitoring_context_s* ctx);

/* ========================================================================== */
/* Version & Error Handling                                                   */
/* ========================================================================== */

const char* monitoring_version(void) {
    static char version[32];
    snprintf(version, sizeof(version), "%d.%d.%d",
             MONITORING_VERSION_MAJOR,
             MONITORING_VERSION_MINOR,
             MONITORING_VERSION_PATCH);
    return version;
}

const char* monitoring_error_string(monitoring_error_t error) {
    switch (error) {
        case MONITORING_OK: return "Success";
        case MONITORING_ERROR: return "Generic error";
        case MONITORING_INVALID_PARAM: return "Invalid parameter";
        case MONITORING_NOT_INITIALIZED: return "SDK not initialized";
        case MONITORING_ALREADY_INIT: return "Already initialized";
        case MONITORING_NO_MEMORY: return "Out of memory";
        case MONITORING_IO_ERROR: return "I/O error";
        case MONITORING_NETWORK_ERROR: return "Network error";
        case MONITORING_TIMEOUT: return "Timeout";
        case MONITORING_NOT_SUPPORTED: return "Operation not supported";
        default: return "Unknown error";
    }
}

/* ========================================================================== */
/* Configuration & Initialization                                             */
/* ========================================================================== */

monitoring_error_t monitoring_init(const monitoring_config_t* config) {
    if (!config) {
        return MONITORING_INVALID_PARAM;
    }
    
    pthread_mutex_lock(&g_sdk.lock);
    
    if (g_sdk.initialized) {
        pthread_mutex_unlock(&g_sdk.lock);
        return MONITORING_ALREADY_INIT;
    }
    
    /* Copy configuration */
    memcpy(&g_sdk.config, config, sizeof(monitoring_config_t));
    
    /* Initialize backends based on mode */
    if (config->mode == MONITORING_MODE_SIDECAR) {
        /* Sidecar mode: single backend */
        g_sdk.num_backends = 1;
        g_sdk.backends = calloc(1, sizeof(backend_t));
        if (!g_sdk.backends) {
            pthread_mutex_unlock(&g_sdk.lock);
            return MONITORING_NO_MEMORY;
        }
        
        g_sdk.backends[0].type = MONITORING_BACKEND_SIDECAR;
        g_sdk.backends[0].enabled = true;
        g_sdk.backends[0].priority = 1;
        /* TODO: Initialize sidecar backend implementation */
        
    } else {
        /* Direct mode: multiple backends */
        if (config->num_backends > 0) {
            g_sdk.num_backends = config->num_backends;
            g_sdk.backends = calloc(config->num_backends, sizeof(backend_t));
            if (!g_sdk.backends) {
                pthread_mutex_unlock(&g_sdk.lock);
                return MONITORING_NO_MEMORY;
            }
            
            for (int i = 0; i < config->num_backends; i++) {
                g_sdk.backends[i].type = config->backends[i].type;
                g_sdk.backends[i].enabled = config->backends[i].enabled;
                g_sdk.backends[i].priority = config->backends[i].priority;
                /* TODO: Initialize backend implementations */
            }
        }
    }
    
    g_sdk.initialized = true;
    pthread_mutex_unlock(&g_sdk.lock);
    
    return MONITORING_OK;
}

monitoring_error_t monitoring_init_from_file(const char* config_file) {
    if (!config_file) {
        return MONITORING_INVALID_PARAM;
    }
    
    /* TODO: Parse JSON configuration file */
    return MONITORING_NOT_SUPPORTED;
}

monitoring_error_t monitoring_shutdown(void) {
    pthread_mutex_lock(&g_sdk.lock);
    
    if (!g_sdk.initialized) {
        pthread_mutex_unlock(&g_sdk.lock);
        return MONITORING_NOT_INITIALIZED;
    }
    
    /* Cleanup backends */
    if (g_sdk.backends) {
        /* TODO: Close backend implementations */
        free(g_sdk.backends);
        g_sdk.backends = NULL;
    }
    
    g_sdk.num_backends = 0;
    g_sdk.initialized = false;
    
    pthread_mutex_unlock(&g_sdk.lock);
    
    return MONITORING_OK;
}

bool monitoring_is_initialized(void) {
    bool result;
    pthread_mutex_lock(&g_sdk.lock);
    result = g_sdk.initialized;
    pthread_mutex_unlock(&g_sdk.lock);
    return result;
}

/* ========================================================================== */
/* Event API                                                                  */
/* ========================================================================== */

monitoring_error_t monitoring_send_event(const monitoring_event_t* event) {
    if (!event) {
        return MONITORING_INVALID_PARAM;
    }
    
    if (!monitoring_is_initialized()) {
        return MONITORING_NOT_INITIALIZED;
    }
    
    return send_event_to_backends(event);
}

monitoring_error_t monitoring_send_batch(const monitoring_event_t* events, int count) {
    if (!events || count <= 0) {
        return MONITORING_INVALID_PARAM;
    }
    
    if (!monitoring_is_initialized()) {
        return MONITORING_NOT_INITIALIZED;
    }
    
    /* Send each event individually for now */
    /* TODO: Implement true batch support */
    for (int i = 0; i < count; i++) {
        monitoring_error_t err = send_event_to_backends(&events[i]);
        if (err != MONITORING_OK) {
            return err;
        }
    }
    
    return MONITORING_OK;
}

static monitoring_error_t send_event_to_backends(const monitoring_event_t* event) {
    /* TODO: Send to all enabled backends */
    /* For now, just log */
    fprintf(stderr, "[monitoring] Event: %s (kind=%d)\n", 
            event->idempotency_key, event->event_kind);
    return MONITORING_OK;
}

/* ========================================================================== */
/* Context API                                                                */
/* ========================================================================== */

monitoring_context_t monitoring_start(const char* name, const char* entity_id) {
    if (!name || !entity_id) {
        return NULL;
    }
    
    if (!monitoring_is_initialized()) {
        return NULL;
    }
    
    monitoring_context_s* ctx = calloc(1, sizeof(monitoring_context_s));
    if (!ctx) {
        return NULL;
    }
    
    ctx->name = strdup(name);
    ctx->entity_id = strdup(entity_id);
    ctx->entity_type = MONITORING_ENTITY_JOB;
    ctx->start_time = time(NULL);
    
    /* Allocate initial capacity for metrics/metadata */
    ctx->metrics_capacity = 16;
    ctx->metric_keys = calloc(ctx->metrics_capacity, sizeof(char*));
    ctx->metric_values = calloc(ctx->metrics_capacity, sizeof(double));
    
    ctx->metadata_capacity = 16;
    ctx->metadata_keys = calloc(ctx->metadata_capacity, sizeof(char*));
    ctx->metadata_values = calloc(ctx->metadata_capacity, sizeof(char*));
    
    /* Send started event */
    char event_id[64];
    snprintf(event_id, sizeof(event_id), "%s-start-%ld", entity_id, ctx->start_time);
    
    monitoring_event_t event = {
        .idempotency_key = event_id,
        .site_id = g_sdk.config.site_id,
        .app_name = g_sdk.config.app_name,
        .app_version = g_sdk.config.app_version,
        .entity_type = ctx->entity_type,
        .entity_id = entity_id,
        .event_kind = MONITORING_EVENT_STARTED,
        .timestamp = ctx->start_time,
        .status = "started",
        .num_metrics = 0,
        .num_metadata = 0
    };
    
    monitoring_send_event(&event);
    
    return ctx;
}

monitoring_error_t monitoring_progress(monitoring_context_t ctx, int progress, const char* message) {
    if (!ctx) {
        return MONITORING_INVALID_PARAM;
    }
    
    monitoring_context_s* context = (monitoring_context_s*)ctx;
    
    char event_id[64];
    snprintf(event_id, sizeof(event_id), "%s-progress-%ld", 
             context->entity_id, time(NULL));
    
    /* Create progress metric */
    const char* keys[] = {"progress"};
    double values[] = {(double)progress};
    
    monitoring_event_t event = {
        .idempotency_key = event_id,
        .site_id = g_sdk.config.site_id,
        .app_name = g_sdk.config.app_name,
        .app_version = g_sdk.config.app_version,
        .entity_type = context->entity_type,
        .entity_id = context->entity_id,
        .event_kind = MONITORING_EVENT_PROGRESS,
        .timestamp = time(NULL),
        .status = message ? message : "in_progress",
        .metric_keys = keys,
        .metric_values = values,
        .num_metrics = 1,
        .num_metadata = 0
    };
    
    return monitoring_send_event(&event);
}

monitoring_error_t monitoring_add_metric(monitoring_context_t ctx, const char* key, double value) {
    if (!ctx || !key) {
        return MONITORING_INVALID_PARAM;
    }
    
    monitoring_context_s* context = (monitoring_context_s*)ctx;
    
    /* Grow array if needed */
    if (context->num_metrics >= context->metrics_capacity) {
        context->metrics_capacity *= 2;
        context->metric_keys = realloc(context->metric_keys, 
                                      context->metrics_capacity * sizeof(char*));
        context->metric_values = realloc(context->metric_values,
                                        context->metrics_capacity * sizeof(double));
    }
    
    context->metric_keys[context->num_metrics] = strdup(key);
    context->metric_values[context->num_metrics] = value;
    context->num_metrics++;
    
    return MONITORING_OK;
}

monitoring_error_t monitoring_add_metadata(monitoring_context_t ctx, const char* key, const char* value) {
    if (!ctx || !key || !value) {
        return MONITORING_INVALID_PARAM;
    }
    
    monitoring_context_s* context = (monitoring_context_s*)ctx;
    
    /* Grow array if needed */
    if (context->num_metadata >= context->metadata_capacity) {
        context->metadata_capacity *= 2;
        context->metadata_keys = realloc(context->metadata_keys,
                                        context->metadata_capacity * sizeof(char*));
        context->metadata_values = realloc(context->metadata_values,
                                          context->metadata_capacity * sizeof(char*));
    }
    
    context->metadata_keys[context->num_metadata] = strdup(key);
    context->metadata_values[context->num_metadata] = strdup(value);
    context->num_metadata++;
    
    return MONITORING_OK;
}

monitoring_error_t monitoring_finish(monitoring_context_t ctx) {
    if (!ctx) {
        return MONITORING_INVALID_PARAM;
    }
    
    monitoring_context_s* context = (monitoring_context_s*)ctx;
    
    char event_id[64];
    time_t now = time(NULL);
    snprintf(event_id, sizeof(event_id), "%s-finish-%ld", context->entity_id, now);
    
    /* Add duration metric */
    double duration = difftime(now, context->start_time);
    monitoring_add_metric(ctx, "duration_seconds", duration);
    
    monitoring_event_t event = {
        .idempotency_key = event_id,
        .site_id = g_sdk.config.site_id,
        .app_name = g_sdk.config.app_name,
        .app_version = g_sdk.config.app_version,
        .entity_type = context->entity_type,
        .entity_id = context->entity_id,
        .event_kind = MONITORING_EVENT_FINISHED,
        .timestamp = now,
        .status = "success",
        .metric_keys = (const char**)context->metric_keys,
        .metric_values = context->metric_values,
        .num_metrics = context->num_metrics,
        .metadata_keys = (const char**)context->metadata_keys,
        .metadata_values = (const char**)context->metadata_values,
        .num_metadata = context->num_metadata
    };
    
    monitoring_error_t err = monitoring_send_event(&event);
    free_context_internal(context);
    
    return err;
}

monitoring_error_t monitoring_error(monitoring_context_t ctx, const char* error_message) {
    if (!ctx) {
        return MONITORING_INVALID_PARAM;
    }
    
    monitoring_context_s* context = (monitoring_context_s*)ctx;
    
    /* Add error as metadata */
    if (error_message) {
        monitoring_add_metadata(ctx, "error", error_message);
    }
    
    char event_id[64];
    snprintf(event_id, sizeof(event_id), "%s-error-%ld", 
             context->entity_id, time(NULL));
    
    monitoring_event_t event = {
        .idempotency_key = event_id,
        .site_id = g_sdk.config.site_id,
        .app_name = g_sdk.config.app_name,
        .app_version = g_sdk.config.app_version,
        .entity_type = context->entity_type,
        .entity_id = context->entity_id,
        .event_kind = MONITORING_EVENT_ERROR,
        .timestamp = time(NULL),
        .status = "error",
        .metric_keys = (const char**)context->metric_keys,
        .metric_values = context->metric_values,
        .num_metrics = context->num_metrics,
        .metadata_keys = (const char**)context->metadata_keys,
        .metadata_values = (const char**)context->metadata_values,
        .num_metadata = context->num_metadata
    };
    
    monitoring_error_t err = monitoring_send_event(&event);
    free_context_internal(context);
    
    return err;
}

monitoring_error_t monitoring_cancel(monitoring_context_t ctx) {
    if (!ctx) {
        return MONITORING_INVALID_PARAM;
    }
    
    monitoring_context_s* context = (monitoring_context_s*)ctx;
    
    char event_id[64];
    snprintf(event_id, sizeof(event_id), "%s-cancel-%ld",
             context->entity_id, time(NULL));
    
    monitoring_event_t event = {
        .idempotency_key = event_id,
        .site_id = g_sdk.config.site_id,
        .app_name = g_sdk.config.app_name,
        .app_version = g_sdk.config.app_version,
        .entity_type = context->entity_type,
        .entity_id = context->entity_id,
        .event_kind = MONITORING_EVENT_CANCELED,
        .timestamp = time(NULL),
        .status = "canceled",
        .num_metrics = 0,
        .num_metadata = 0
    };
    
    monitoring_error_t err = monitoring_send_event(&event);
    free_context_internal(context);
    
    return err;
}

static void free_context_internal(monitoring_context_s* ctx) {
    if (!ctx) return;
    
    free(ctx->name);
    free(ctx->entity_id);
    
    /* Free metrics */
    for (int i = 0; i < ctx->num_metrics; i++) {
        free(ctx->metric_keys[i]);
    }
    free(ctx->metric_keys);
    free(ctx->metric_values);
    
    /* Free metadata */
    for (int i = 0; i < ctx->num_metadata; i++) {
        free(ctx->metadata_keys[i]);
        free(ctx->metadata_values[i]);
    }
    free(ctx->metadata_keys);
    free(ctx->metadata_values);
    
    free(ctx);
}

/* ========================================================================== */
/* Utilities                                                                  */
/* ========================================================================== */

char* monitoring_generate_id(char* buffer) {
    if (!buffer) {
        return NULL;
    }
    
#ifdef __APPLE__
    uuid_t uuid;
    uuid_generate(uuid);
    uuid_unparse(uuid, buffer);
#else
    /* Fallback: use timestamp + random */
    snprintf(buffer, 37, "%08x-%04x-%04x-%04x-%012lx",
             (unsigned int)time(NULL),
             rand() % 0xFFFF,
             rand() % 0xFFFF,
             rand() % 0xFFFF,
             (unsigned long)rand() % 0xFFFFFFFFFFFF);
#endif
    
    return buffer;
}

time_t monitoring_timestamp(void) {
    return time(NULL);
}

monitoring_error_t monitoring_health_check(void) {
    if (!monitoring_is_initialized()) {
        return MONITORING_NOT_INITIALIZED;
    }
    
    /* TODO: Check backend health */
    return MONITORING_OK;
}

